opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py

```diff
@@ -0,0 +1,241 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Sequence, Tuple, Type
+
+from opentelemetry import metrics as metrics_api
+
+logger = logging.getLogger(__name__)
+
+
+class BaseHandle:
+    def __init__(
+        self,
+        value_type: Type[metrics_api.ValueT],
+        enabled: bool,
+        monotonic: bool,
+    ):
+        self.data = 0
+        self.value_type = value_type
+        self.enabled = enabled
+        self.monotonic = monotonic
+
+    def _validate_update(self, value: metrics_api.ValueT):
+        if not self.enabled:
+            return False
+        if self.monotonic and value < 0:
+            logger.warning("Monotonic metric cannot descend.")
+            return False
+        if not isinstance(value, self.value_type):
+            logger.warning(
+                "Invalid value passed for %s.", self.value_type.__name__
+            )
+            return False
+        return True
+
+
+class CounterHandle(metrics_api.CounterHandle, BaseHandle):
+    def update(self, value: metrics_api.ValueT) -> None:
+        if self._validate_update(value):
+            self.data += value
+
+    def add(self, value: metrics_api.ValueT) -> None:
+        """See `opentelemetry.metrics.CounterHandle._add`."""
+        self.update(value)
+
+
+class GaugeHandle(metrics_api.GaugeHandle, BaseHandle):
+    def update(self, value: metrics_api.ValueT) -> None:
+        if self._validate_update(value):
+            self.data = value
+
+    def set(self, value: metrics_api.ValueT) -> None:
+        """See `opentelemetry.metrics.GaugeHandle._set`."""
+        self.update(value)
+
+
+class MeasureHandle(metrics_api.MeasureHandle, BaseHandle):
+    def update(self, value: metrics_api.ValueT) -> None:
+        if self._validate_update(value):
+            pass
+            # TODO: record
+
+    def record(self, value: metrics_api.ValueT) -> None:
+        """See `opentelemetry.metrics.MeasureHandle._record`."""
+        self.update(value)
+
+
+class Metric(metrics_api.Metric):
+    """See `opentelemetry.metrics.Metric`."""
+
+    HANDLE_TYPE = BaseHandle
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        unit: str,
+        value_type: Type[metrics_api.ValueT],
+        label_keys: Sequence[str] = None,
+        enabled: bool = True,
+        monotonic: bool = False,
+    ):
+        self.name = name
+        self.description = description
+        self.unit = unit
+        self.value_type = value_type
+        self.label_keys = label_keys
+        self.enabled = enabled
+        self.monotonic = monotonic
+        self.handles = {}
+
+    def get_handle(self, label_values: Sequence[str]) -> BaseHandle:
+        """See `opentelemetry.metrics.Metric.get_handle`."""
+        handle = self.handles.get(label_values)
+        if not handle:
+            handle = self.__class__.HANDLE_TYPE(
```

Comment:
```suggestion
            handle = self.HANDLE_TYPE(
```
Nit: you can get a class attr by name if there's not an instance attr shadowing it.